### PR TITLE
Add ModifyElementSubdomainByVar UserObject

### PR DESCRIPTION
**Pull request by dewenyushu**
Address #27486

### CODE REVIEW

File: framework/include/userobjects/ModifyElementSubdomainByVar.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ElementSubdomainModifier.h"
+#include "MooseEnum.h"
+
+class ModifyElementSubdomainByVar : public ElementSubdomainModifier
```
Comment (GiudGiud) [inline framework/include/userobjects/ModifyElementSubdomainByVar.h:15]: docstring

File: framework/src/userobjects/ModifyElementSubdomainByVar.C [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  : ElementSubdomainModifier(parameters), _v(coupledValue("coupled_var"))
+{
+}
+
+SubdomainID
+ModifyElementSubdomainByVar::computeSubdomainID()
+{
+  // Calculate the desired subdomain ID for the current element.
+  Real val = 0.0;
+  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
```
Comment (GiudGiud) [inline framework/src/userobjects/ModifyElementSubdomainByVar.C:36]:
```suggestion
  for (const auto qp : make_range(_qrule->n_points()))
```

File: framework/src/userobjects/ModifyElementSubdomainByVar.C [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  SubdomainID sid = (unsigned int)(round(val));
+
+  // Check if the desired subdomain ID exists in the mesh. Use the nearest subdomain ID if the
+  // desired subdomain ID does not exist in the mesh.
+  if (_mesh.meshSubdomains().find(sid) == _mesh.meshSubdomains().end())
+  {
+    auto it = _mesh.meshSubdomains().lower_bound(sid);
+    SubdomainID nearest_id =
+        it == _mesh.meshSubdomains().end() ? *_mesh.meshSubdomains().rbegin() : *it;
+    mooseDoOnce(mooseWarning(
```
Comment (GiudGiud) [inline framework/src/userobjects/ModifyElementSubdomainByVar.C:49]: We dont like to do this anymore.

I suggest you keep a vector (with a lock for threads?) that keeps track of which subdomains you have output a warning for, to output the warning only once per subdomain

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: framework/src/userobjects/ModifyElementSubdomainByVar.C [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    {
+      mooseWarning("Requested subdomain ",
+                   sid,
+                   " does not exist. Subdomain ID ",
+                   nearest_id,
+                   " is assigned. Please ensure the passed variable falls within the range of the "
+                   "mesh's subdomain IDs for the expected behavior.");
+      _void_sids.insert(sid);
+    }
+    return nearest_id;
```
Comment (dewenyushu) [inline framework/src/userobjects/ModifyElementSubdomainByVar.C:63]: @GiudGiud thanks for the comment about adding a lock for thread-safe operations. I am guessing this is the correct way of doing it?

Comment (GiudGiud) [reply framework/src/userobjects/ModifyElementSubdomainByVar.C:63]: looks good to me

**Comment by dewenyushu** [comment]
@GiudGiud this should be ready :)

File: framework/src/userobjects/ModifyElementSubdomainByVar.C [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+  SubdomainID sid = (unsigned int)(round(val));
+
+  // Check if the target subdomain ID exists in the mesh. Use the nearest subdomain ID if the
+  // target subdomain ID does not exist in the mesh.
+  if (_mesh.meshSubdomains().find(sid) == _mesh.meshSubdomains().end())
+  {
+    auto it = _mesh.meshSubdomains().lower_bound(sid);
+    SubdomainID nearest_id =
+        it == _mesh.meshSubdomains().end() ? *_mesh.meshSubdomains().rbegin() : *it;
```
Comment (GiudGiud) [inline framework/src/userobjects/ModifyElementSubdomainByVar.C:48]: this isnt the nearest in a rounding sense? 
https://cplusplus.com/reference/set/set/lower_bound/
This is the one right below, with an adaptation if the end is found (so sid is bigger than everything?)

either change the wording in the doco to describe in what sense you mean closest, or compare to the next element to find the closest

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: framework/src/userobjects/ModifyElementSubdomainByVar.C (hunk 5 above)
Comment (dewenyushu) [reply framework/src/userobjects/ModifyElementSubdomainByVar.C]: Yeah good point. The lower bound does not equal to closest. It returns the pointer to the smallest value that is equal or greater than the inquired value. I just changed the naming and the doc to reflect this

**Comment by dewenyushu** [comment]
@GiudGiud this should be ready (again)

**Comment by hugary1995** [comment]
I'd suggest a different name. To have slightly more consistency with the existing `CoupledVarThresholdElementSubdomainModifier`, this one could be named `VariableValueElementSubdomainModifier` which might also be more self-explanatory.

**Comment by dewenyushu** [comment]
> I'd suggest a different name. To have slightly more consistency with the existing `CoupledVarThresholdElementSubdomainModifier`, this one could be named `VariableValueElementSubdomainModifier` which might also be more self-explanatory.

Name changed!

File: framework/src/userobjects/VariableValueElementSubdomainModifier.C [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  for (const auto qp : make_range(_qrule->n_points()))
+    val += _v[qp] * _JxW[qp] * _coord[qp];
+  val /= _current_elem_volume;
+
+  SubdomainID sid = (unsigned int)(round(val));
+
+  // Verify whether the specified target subdomain ID is present within the mesh. If it is not
+  // found, locate the smallest subdomain ID in the mesh that matches or exceeds the target
+  // subdomain ID. Or select the largest subdomain ID present in the mesh if all subdomain IDs are
+  // smaller than the target.
```
Comment (GiudGiud) [inline framework/src/userobjects/VariableValueElementSubdomainModifier.C:46]: basically this but in the doco

File: framework/doc/content/source/userobjects/ModifyElementSubdomainByVar.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# ModifyElementSubdomainByVar
+
+The `ModifyElementSubdomainByVar` UserObject adjusts the element's subdomain according to a provided variable value. Subdomain IDs can be assigned based on the average variable value within the element, rounded to the nearest existing subdomain ID in the mesh.
+
+**Note:** If the desired value is not present in the subdomain ID list, a warning will be issued, and the closest existing subdomain ID value will be automatically assigned to the element.
```
[the remark this answers is not in the document]
Comment (dewenyushu) [reply framework/doc/content/source/userobjects/ModifyElementSubdomainByVar.md]: Fixed!

**Review by GiudGiud — APPROVED** [review]

File: test/tests/userobjects/element_subdomain_modifier/tests [hunk 8]
```diff
@@ -1,6 +1,6 @@
-[Tests]
-  issues = '#15795 #17100 #17239 #21383'
-  design = 'CoupledVarThresholdElementSubdomainModifier.md'
+ [Tests]
+  issues = '#15795 #17100 #17239 #21383 #27486'
+  design = 'CoupledVarThresholdElementSubdomainModifier.md ModifyElementSubdomainByVar.md'
```
Comment (GiudGiud) [inline test/tests/userobjects/element_subdomain_modifier/tests:3]:
```suggestion
  design = 'CoupledVarThresholdElementSubdomainModifier.md VariableValueElementSubdomainModifier.md'
```

**Pull request merged**